runtime/compiler/codegen/PrivateLinkage.cpp

```diff
@@ -0,0 +1,40 @@
+/*******************************************************************************
+ * Copyright (c) 2020, 2020 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "codegen/CodeGenerator.hpp"
+#include "codegen/Linkage_inlines.hpp"
+#include "codegen/PrivateLinkage.hpp"
+#include "env/jittypes.h"
+
+intptr_t
+J9::PrivateLinkage::entryPointFromCompiledMethod()
+   {
+   uint8_t *methodEntry = cg()->getCodeStart();
+   methodEntry += J9::PrivateLinkage::LinkageInfo::get(methodEntry)->getReservedWord();
+   return reinterpret_cast<intptr_t>(methodEntry);
+   }
+
+intptr_t
+J9::PrivateLinkage::entryPointFromInterpretedMethod()
+   {
+   return reinterpret_cast<intptr_t>(cg()->getCodeStart());
```

Comment:
IIRC in JITServer mode, for compilations that happen remotely calling `getCodeStart()` on the client's side doesn't return anything meaningful because the client's code generator instance isn't used. That would mean that `entryPointFromInterpretedMethod()` can't be used in some contexts. Is that a problem?

I just happen to recall this because I fixed a similar bug a while back by getting the same info from the metadata instead: https://github.com/eclipse/openj9/commit/cc3f690b29d06fa2ebcc2beb331a79f71dda851f?w=1

Reply:
The intention is for these APIs is to answer in the context of (and within the confines of) the current compilation.  That is, during code generation when you have a Linkage object in hand.  They should return 0 if the address isn't known yet.

They obtain their information in the same way that is done presently, just hiding it behind an API.  Does JitServer have to do something special when you ask for `cg->getCodeStart()` during a compilation now?  If the answer is no then I think these APIs are fine (I sniffed a couple of uses of `getCodeStart()` but didn't see anything special for JitServer).  But if yes, then you're right they may need to handle JitServer a bit differently.

Reply:
> The intention is for these APIs is to answer in the context of (and within the confines of) the current compilation. That is, during code generation when you have a Linkage object in hand. They should return 0 if the address isn't known yet.

OK, within the confines of the code generator it should work as expected.

> Does JitServer have to do something special when you ask for cg->getCodeStart() during a compilation now?

Yes, depending on where (either the client or the server) the ask is happening:

1. If the code making the call is executing on the server it works fine (the codegen phase runs on the server so asking during Instruction Selection, Binary Encoding, etc works).
2. If the code is executing on the client it always returns `0`, even if the address is known.

We've fixed the instances of 2 by not calling `getCodeStart()` and instead finding the address via other means (for example via the metadata).

My point was that `entryPointFromInterpretedMethod()` is going to have the same limitation, so if you have plans to use it outside of the codegen phase you might run into issues down the road.

Reply:
These APIs are only used for introspection on the current method being compiled as it is being compiled (i.e., within the confines of the compilation).  They are supposed to return "0" if the code start address is not known.  There isn't an alternate mechanism to fall back on during the compilation, so either the code start address is known or it returns 0.  I imagine any code calling this API is going to have to deal with the "0" case and find an alternate means to get the information it needs.